dune/grid/polyhedralgrid/grid.hh

```diff
@@ -0,0 +1,1207 @@
+    {
+      static std::vector< GeometryType > emptyDummy;
+      if (0 <= codim && codim < geomTypes_.size())
+        return geomTypes_[codim];
+
+      return emptyDummy;
+    }
+
+    int indexInInside( const typename Codim<0>::EntitySeed& seed, const int i ) const
+    {
+      if( ! grid_.cell_facetag )
+        return i;
+      else
+      {
+        // assert( i>= 0 && i<subEntities( EntitySeed( seed.index() ) ) );
+        return grid_.cell_facetag[ grid_.cell_facepos[ seed.index() ] + i ] ;
+      }
+    }
+
+    typename Codim<0>::EntitySeed
+    neighbor( const typename Codim<0>::EntitySeed& seed, const int i ) const
+    {
+      typedef typename Codim<0>::EntitySeed EntitySeed;
+      const int face = 2 * this->template subEntitySeed<1>( seed, i ).index();
+      int nb = grid_.face_cells[ face ];
+      if( nb == seed.index() )
+        nb = grid_.face_cells[ face+1 ];
+
+      return EntitySeed( nb );
+    }
+
+    int
+    indexInOutside( const typename Codim<0>::EntitySeed& seed, const int i ) const
+    {
+      typedef typename Codim<0>::EntitySeed EntitySeed;
+      EntitySeed nb = neighbor( seed, i );
+      const int faces = subEntities( seed, 1 );
+      for( int face = 0; face<faces; ++ face )
+      {
+        if( neighbor( nb, face ).equals(seed) )
+          return indexInInside( nb, face );
+      }
+      DUNE_THROW(InvalidStateException,"inverse intersection not found");
+      return -1;
+    }
+
+    template <class EntitySeed>
+    GlobalCoordinate
+    outerNormal( const EntitySeed& seed, const int i ) const
+    {
+      const int face  = this->template subEntitySeed<1>( seed, i ).index();
+      const int normalIdx = face * GlobalCoordinate :: dimension ;
+      GlobalCoordinate normal = copyToGlobalCoordinate( grid_.face_normals + normalIdx );
+      const int nb = grid_.face_cells[ face ];
+      if( nb != seed.index() )
+        normal *= -1.0;
+      return std::move( normal );
+    }
+
+    template <class EntitySeed>
+    GlobalCoordinate centroids( const EntitySeed& seed ) const
+    {
+      const int index = GlobalCoordinate :: dimension * seed.index();
+      const int codim = EntitySeed::codimension;
+      assert( index >= 0 && index < size( codim ) * GlobalCoordinate :: dimension );
+
+      if( codim == 0 )
+        return copyToGlobalCoordinate( grid_.cell_centroids + index );
+      else if ( codim == 1 )
+        return copyToGlobalCoordinate( grid_.face_centroids + index );
+      else if( codim == dim )
+        return copyToGlobalCoordinate( grid_.node_coordinates + index );
+      else
+      {
+        DUNE_THROW(InvalidStateException,"codimension not implemented");
+        return GlobalCoordinate( 0 );
+      }
+    }
+
+    GlobalCoordinate copyToGlobalCoordinate( const double* coords ) const
+    {
+      GlobalCoordinate coordinate;
+      for( int i=0; i<GlobalCoordinate::dimension; ++i )
+        coordinate[ i ] = coords[ i ];
+      return std::move( coordinate );
```

Comment:
Since GlobalCoordinate is a Dune::FieldVector (it is, right?) I do not think move is going to help since a move is the same as a copy for that class. I do not think it hurts either (by preventing the return value optimization, which I think *is* important here) although I am not 100% sure.